### DISCUSSION TITLE
built in Nextjs caching doesn't work with custom function

### DISCUSSION THREAD

**Discussion by AliMilani13** [vercel/next.js#70697] [Help]
### Summary

Hello everyone. 

I've noticed that, if you use `fetch` by itself, it works as expected and caches the response. but when I tried to implement something like the one in the code, it sends the request to the server, every time I refresh the page. that means caching magic doesn't work in this case.  which makes me wonder why this is the case? I'm still using fetch API with custom parameters and error handlings. 

I would be thankful if you help. 

### Additional information

```js
import { cookies } from "next/headers"
import { _baseUrl } from "../_baseurl/_baseurl";



/**
 * custom fetch request with get method
 * @param {string} url API address coming from .env file
 * @param {bool} auth authorization. will set bearer token to header if it is true
 */
export const req_get= async(url,auth=false)=>{
  const __accessToken = cookies().get('access-token');
  
  try {
    const response  = await fetch(`${_baseUrl}${url}`,{
        method:'get',
        ...(auth?{
            headers:{
                Authorization: `Bearer ${__accessToken}`
            }
        }:{}),
    
      });
  
  
    if (!response.ok) {
      throw new Error('Network response was not ok');
    }
 
    const data = await response.json();

   return data;
  } catch (error) {
    // Handle any errors here
   return  console.error(`Server error at API endpoint:${url}`, error);
  }

}
```

**Comment by suprunchuk** [comment]
The reason why caching might not be working as expected in your custom fetch implementation is due to a few factors:

Default caching behavior: By default, fetch uses a "no-store" cache policy, which means it doesn't cache responses unless explicitly told to do so.

Custom headers: When you add custom headers (like the Authorization header in your case), it can affect the caching behavior. Requests with different headers are considered different requests, even if the URL is the same.

Server-side rendering: If this code is running on the server-side (which seems likely given the use of cookies() from Next.js), caching behavior can be different compared to client-side fetching.

To enable caching for your custom fetch implementation, you can try the following modifications:

```tsx

import { cookies } from "next/headers"
import { _baseUrl } from "../_baseurl/_baseurl";

export const req_get = async (url, auth = false) => {
  const __accessToken = cookies().get('access-token');
  
  try {
    const response = await fetch(`${_baseUrl}${url}`, {
      method: 'GET',
      ...(auth ? {
        headers: {
          Authorization: `Bearer ${__accessToken}`
        }
      } : {}),
      cache: 'force-cache', // Enable caching
      next: { revalidate: 3600 }, // Revalidate every hour (3600 seconds)
    });

    if (!response.ok) {
      throw new Error('Network response was not ok');
    }
 
    const data = await response.json();
    return data;
  } catch (error) {
    console.error(`Server error at API endpoint:${url}`, error);
    throw error; // Re-throw the error instead of returning it
  }
}

```
